### ISSUE TITLE
change bucket_cap_mb in DistributedDataParallel and md5 of grad change

### ISSUE THREAD

**Issue by 522730312** [pytorch/pytorch#30509]
## 🐛 Bug
when I use funtion as below and change bucket_cap_mb form 20 to 25, the md5 of my grad has change,is there any one can explain about this?

model = torch.nn.parallel.DistributedDataParallel(
        model,
        device_ids=[args.device_id],
        output_device=args.device_id,
        bucket_cap_mb=20)
 
The MD5 of different bucket_cap_mb:
19:9217933d100373decf50026c551d3f72
20:9217933d100373decf50026c551d3f72
21:977a9c79b96acf3f095634ad94a6a681
22:977a9c79b96acf3f095634ad94a6a681
23:977a9c79b96acf3f095634ad94a6a681
24:977a9c79b96acf3f095634ad94a6a681

## Expected behavior
    The MD5 of Grad should stable

## Environment
     GPU:V100*4  32G
     pytorch:1.2.0
     OS: Red-hat
     gcc:4.8.5

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528

**Comment by pietern** [comment]
It is possible that NCCL chooses to use a different algorithm when operating on memory larger than a certain threshold. Floating point addition is not associative, so changing the order of addition can mean a slightly different outcome.

To cover our bases, can you check if the norm of those grads is the same?

**Comment by pietern** [comment]
Which model can you reproduce this with?

**Comment by 522730312** [comment]
Here is my model，and run with three commands:
python ddp_mnist.py --rank 0 --world-size 3
python ddp_mnist.py --rank 1 --world-size 3
python ddp_mnist.py --rank 2 --world-size 3

`#!/usr/bin/env python
# coding=utf-8
from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel
import torch.nn.functional as F

import numpy as np
import os

# Training settings
parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
parser.add_argument('--rank', type=int)
parser.add_argument('--world-size',type=int, default=3)
parser.add_argument('--cuda',type=bool, default=True)
parser.add_argument('--init-method', type=str, default='tcp://127.0.0.1:23456')

args = parser.parse_args()
args.cuda = True
os.environ["CUDA_VISIBLE_DEVICES"] = str(args.rank)

dist.init_process_group(
                        init_method=args.init_method,
                        backend="nccl",
                        world_size=args.world_size,
                        rank=args.rank,
                        group_name="pytorch_test")

torch.manual_seed(611+args.rank)
if args.cuda:
    torch.cuda.manual_seed(611+args.rank)


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout2d(0.25)
        self.dropout2 = nn.Dropout2d(0.5)
        self.fc1 = nn.Linear(9216, 1024)
        self.fc2 = nn.Linear(1024, 1024)
        self.fc3 = nn.Linear(1024, 1024)
        self.fc4 = nn.Linear(1024, 512)
        self.fc5 = nn.Linear(512, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        x = self.fc3(x)
        x = self.fc4(x)
        x = self.fc5(x)
        output = F.log_softmax(x, dim=1)
        return output

model = Net()


device = torch.device('cuda', args.rank % torch.cuda.device_count())
model.to(device)
ce_criterion = nn.CrossEntropyLoss().to(device)

model = torch.nn.parallel.DistributedDataParallel(
            model,
            device_ids=[0],
            bucket_cap_mb=30)

optimizer = torch.optim.SGD(
            model.parameters(),
            lr=1e-5,
            momentum=0.1968)

print("this rank:", args.rank)
for i in range(20):
    data = torch.rand(64, 1, 28, 28)
    target = torch.ones(64, dtype=torch.long)

    with torch.cuda.device(device):
        data = data.to(device)
        target = target.to(device)
    optimizer.zero_grad()
    output = model(data)
    loss = ce_criterion(output, target)
    loss.backward()
    optimizer.step()
    print(i)
    if args.rank == 0 and i==10:
        for index,param in enumerate(model.parameters()):
            if hasattr(param, "grad"):
                np.save("tmp/%s.npy"%(index), param.grad.cpu().numpy())`

**Comment by 522730312** [comment]
Here is my all code and ru with three commands below:
python mnist.py --rank 0 --world-size 3
python mnist.py --rank 1 --world-size 3
python mnist.py --rank 2 --world-size 3
`from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel
import torch.nn.functional as F

import numpy as np
import os
parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
parser.add_argument('--rank', type=int)
parser.add_argument('--world-size',type=int, default=3)
parser.add_argument('--cuda',type=bool, default=True)
parser.add_argument('--init-method', type=str, default='tcp://127.0.0.1:23456')

args = parser.parse_args()
args.cuda = True
os.environ["CUDA_VISIBLE_DEVICES"] = str(args.rank)

dist.init_process_group(
                        init_method=args.init_method,
                        backend="nccl",
                        world_size=args.world_size,
                        rank=args.rank,
                        group_name="pytorch_test")

torch.manual_seed(611+args.rank)
if args.cuda:
    torch.cuda.manual_seed(611+args.rank)


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout2d(0.25)
        self.dropout2 = nn.Dropout2d(0.5)
        self.fc1 = nn.Linear(9216, 1024)
        self.fc2 = nn.Linear(1024, 1024)
        self.fc3 = nn.Linear(1024, 1024)
        self.fc4 = nn.Linear(1024, 512)
        self.fc5 = nn.Linear(512, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        x = self.fc3(x)
        x = self.fc4(x)
        x = self.fc5(x)
        output = F.log_softmax(x, dim=1)
        return output

model = Net()


device = torch.device('cuda', args.rank % torch.cuda.device_count())
model.to(device)
ce_criterion = nn.CrossEntropyLoss().to(device)

model = torch.nn.parallel.DistributedDataParallel(
            model,
            device_ids=[0],
            bucket_cap_mb=30)

optimizer = torch.optim.SGD(
            model.parameters(),
            lr=1e-5,
            momentum=0.1968)

print("this rank:", args.rank)
for i in range(20):
    data = torch.rand(64, 1, 28, 28)
    target = torch.ones(64, dtype=torch.long)

    with torch.cuda.device(device):
        data = data.to(device)
        target = target.to(device)
    optimizer.zero_grad()
    output = model(data)
    loss = ce_criterion(output, target)
    loss.backward()
    optimizer.step()
    print(i)
    if args.rank == 0 and i==10:
        for index,param in enumerate(model.parameters()):
            if hasattr(param, "grad"):
                np.save("tmp/%s.npy"%(index), param.grad.cpu().numpy())

`

**Comment by 522730312** [comment]
Here is my all code and ru with three commands below:
python mnist.py --rank 0 --world-size 3
python mnist.py --rank 1 --world-size 3
python mnist.py --rank 2 --world-size 3


from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel
import torch.nn.functional as F

import numpy as np
import os

parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
parser.add_argument('--rank', type=int)
parser.add_argument('--world-size',type=int, default=3)
parser.add_argument('--cuda',type=bool, default=True)
parser.add_argument('--init-method', type=str, default='tcp://127.0.0.1:23456')

args = parser.parse_args()
args.cuda = True
os.environ["CUDA_VISIBLE_DEVICES"] = str(args.rank)

dist.init_process_group(
                        init_method=args.init_method,
                        backend="nccl",
                        world_size=args.world_size,
                        rank=args.rank,
                        group_name="pytorch_test")

torch.manual_seed(611+args.rank)
if args.cuda:
    torch.cuda.manual_seed(611+args.rank)


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout2d(0.25)
        self.dropout2 = nn.Dropout2d(0.5)
        self.fc1 = nn.Linear(9216, 1024)
        self.fc2 = nn.Linear(1024, 1024)
        self.fc3 = nn.Linear(1024, 1024)
        self.fc4 = nn.Linear(1024, 512)
        self.fc5 = nn.Linear(512, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        x = self.fc3(x)
        x = self.fc4(x)
        x = self.fc5(x)
        output = F.log_softmax(x, dim=1)
        return output

model = Net()
device = torch.device('cuda', args.rank % torch.cuda.device_count())
model.to(device)
ce_criterion = nn.CrossEntropyLoss().to(device)

model = torch.nn.parallel.DistributedDataParallel(
            model,
            device_ids=[0],
            bucket_cap_mb=30)

optimizer = torch.optim.SGD(
            model.parameters(),
            lr=1e-5,
            momentum=0.1968)

print("this rank:", args.rank)
for i in range(20):
    data = torch.rand(64, 1, 28, 28)
    target = torch.ones(64, dtype=torch.long)

    with torch.cuda.device(device):
        data = data.to(device)
        target = target.to(device)
    optimizer.zero_grad()
    output = model(data)
    loss = ce_criterion(output, target)
    loss.backward()
    optimizer.step()
    print(i)
    if args.rank == 0 and i==10:
        for index,param in enumerate(model.parameters()):
            if hasattr(param, "grad"):
                np.save("tmp/%s.npy"%(index), param.grad.cpu().numpy())

**Comment by 522730312** [comment]
Here is the model to reproduce, set bucket_cap_mb from 3 and 30, MD5 change
`class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout2d(0.25)
        self.dropout2 = nn.Dropout2d(0.5)
        self.fc1 = nn.Linear(9216, 1024)
        self.fc2 = nn.Linear(1024, 1024)
        self.fc3 = nn.Linear(1024, 1024)
        self.fc4 = nn.Linear(1024, 512)
        self.fc5 = nn.Linear(512, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        x = self.fc3(x)
        x = self.fc4(x)
        x = self.fc5(x)
        output = F.log_softmax(x, dim=1)
        return output`

**Comment by 522730312** [comment]
`import argparse
import torch
import torch.nn as nn
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel
import torch.nn.functional as F

import numpy as np
import os

parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
parser.add_argument('--rank', type=int)
parser.add_argument('--world-size',type=int, default=3)
parser.add_argument('--cuda',type=bool, default=True)
parser.add_argument('--init-method', type=str, default='tcp://127.0.0.1:23456')

args = parser.parse_args()
args.cuda = True
os.environ["CUDA_VISIBLE_DEVICES"] = str(args.rank)

dist.init_process_group(
                        init_method=args.init_method,
                        backend="nccl",
                        world_size=args.world_size,
                        rank=args.rank,
                        group_name="pytorch_test")

torch.manual_seed(611+args.rank)
if args.cuda:
    torch.cuda.manual_seed(611+args.rank)


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout2d(0.25)
        self.dropout2 = nn.Dropout2d(0.5)
        self.fc1 = nn.Linear(9216, 1024)
        self.fc2 = nn.Linear(1024, 1024)
        self.fc3 = nn.Linear(1024, 1024)
        self.fc4 = nn.Linear(1024, 512)
        self.fc5 = nn.Linear(512, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        x = self.fc3(x)`

**Comment by 522730312** [comment]
Here is my code and run it with three commands as below:
python ddp_mnist.py --rank 0 --world-size 3
python ddp_mnist.py --rank 1 --world-size 3
python ddp_mnist.py --rank 2 --world-size 3
    
    from __future__ import print_function
    import argparse
    import torch
    import torch.nn as nn
    import torch.distributed as dist
    from torch.nn.parallel import DistributedDataParallel
    import torch.nn.functional as F

    import numpy as np
    import os

    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
    parser.add_argument('--rank', type=int)
    parser.add_argument('--world-size',type=int, default=3)
    parser.add_argument('--cuda',type=bool, default=True)
    parser.add_argument('--init-method', type=str, default='tcp://127.0.0.1:23456')

    args = parser.parse_args()
    args.cuda = True
    os.environ["CUDA_VISIBLE_DEVICES"] = str(args.rank)

    dist.init_process_group(
                            init_method=args.init_method,
                            backend="nccl",
                            world_size=args.world_size,
                            rank=args.rank,
                            group_name="pytorch_test")

    torch.manual_seed(611+args.rank)
    if args.cuda:
        torch.cuda.manual_seed(611+args.rank)


    class Net(nn.Module):
        def __init__(self):
            super(Net, self).__init__()
            self.conv1 = nn.Conv2d(1, 32, 3, 1)
            self.conv2 = nn.Conv2d(32, 64, 3, 1)
            self.dropout1 = nn.Dropout2d(0.25)
            self.dropout2 = nn.Dropout2d(0.5)
            self.fc1 = nn.Linear(9216, 1024)
            self.fc2 = nn.Linear(1024, 1024)
            self.fc3 = nn.Linear(1024, 1024)
            self.fc4 = nn.Linear(1024, 512)
            self.fc5 = nn.Linear(512, 10)

        def forward(self, x):
            x = self.conv1(x)
            x = F.relu(x)
            x = self.conv2(x)
            x = F.max_pool2d(x, 2)
            x = self.dropout1(x)
            x = torch.flatten(x, 1)
            x = self.fc1(x)
            x = F.relu(x)
            x = self.dropout2(x)
            x = self.fc2(x)
            x = self.fc3(x)
            x = self.fc4(x)
            x = self.fc5(x)
            output = F.log_softmax(x, dim=1)
            return output

    model = Net()


    device = torch.device('cuda', args.rank % torch.cuda.device_count())
    model.to(device)
    ce_criterion = nn.CrossEntropyLoss().to(device)

    model = torch.nn.parallel.DistributedDataParallel(
                model,
                device_ids=[0],
                bucket_cap_mb=30)

    optimizer = torch.optim.SGD(
                model.parameters(),
                lr=1e-5,
                momentum=0.1968)

    print("this rank:", args.rank)
    for i in range(20):
        data = torch.rand(64, 1, 28, 28)
        target = torch.ones(64, dtype=torch.long)

        with torch.cuda.device(device):
            data = data.to(device)
            target = target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = ce_criterion(output, target)
        loss.backward()
        optimizer.step()
        print(i)
        if args.rank == 0 and i==10:
            for index,param in enumerate(model.parameters()):
                if hasattr(param, "grad"):
                    np.save("tmp/%s.npy"%(index), param.grad.cpu().numpy())

**Comment by pietern** [comment]
I've been trying to reproduce with your script but am not able.

Note that I added the following lines for reproducibility:

```
torch.backends.cudnn.deterministic = True                                                                                                                                                                                                                      
torch.backends.cudnn.benchmark = False
```

Also see https://pytorch.org/docs/stable/notes/randomness.html.

**Comment by pietern** [comment]
Can you try adding those lines and seeing if it still reproduces for you?

Also please test with both the "gloo" and "nccl" distributed backends, to rule out the NCCL hypothesis.

**Comment by rohan-varma** [comment]
I took a look at this, first, as @pietern mentioned, we'd want to set `torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False` , so that the gradients will be the same across iterations.

I used the following script to check the md5 of the gradients: https://gist.github.com/rohan-varma/03d5180bd86e1b5b6e380711119364c9

I checked this for both the gloo and nccl backends, and I'm getting the same md5.

**Comment by pietern** [comment]
@522730312 Can you try out the suggested fixes and see if you can still repro?

**Comment by rohan-varma** [comment]
Closing due to inactivity
